code/espurna/telnet.ino

```diff
@@ -120,13 +127,93 @@ void _telnetData(unsigned char clientId, void *data, size_t len) {
     #if TERMINAL_SUPPORT
         terminalInject(data, len);
     #endif
+}
+
+void _telnetNotifyConnected(unsigned char i) {
+
+    DEBUG_MSG_P(PSTR("[TELNET] Client #%u connected\n"), i);
+
+    // If there is no terminal support automatically dump info and crash data
+    #if TERMINAL_SUPPORT == 0
+        info();
+        wifiDebug();
+        crashDump();
+        crashClear();
+    #endif
+
+    #ifdef ESPURNA_CORE
+        _telnetClientsAuth[i] = true;
+    #else
+        _telnetClientsAuth[i] = !_telnetAuth;
+        if (_telnetAuth) {
+            if (getAdminPass().length()) {
+                _telnetWrite(i, "Password: ");
+            } else {
+                _telnetClientsAuth[i] = true;
+            }
+        }
+    #endif
+
+    _telnetFirst = true;
+    wifiReconnectCheck();
 
 }
 
-void _telnetNewClient(AsyncClient *client) {
+#if TELNET_SERVER == TELNET_SERVER_WIFISERVER
 
-    if (client->localIP() != WiFi.softAPIP()) {
+void _telnetLoop() {
+    if (_telnetServer.hasClient()) {
+        int i;
+
+        for (i = 0; i < TELNET_MAX_CLIENTS; i++) {
+            if (!_telnetClients[i] || !_telnetClients[i]->connected()) {
+
+                _telnetClients[i] = std::unique_ptr<WiFiClient>(new WiFiClient(_telnetServer.available()));
+
+                if (_telnetClients[i]->localIP() != WiFi.softAPIP()) {
+                    // Telnet is always available for the ESPurna Core image
+                    #ifdef ESPURNA_CORE
+                        bool telnetSTA = true;
+                    #else
+                        bool telnetSTA = getSetting("telnetSTA", TELNET_STA).toInt() == 1;
+                    #endif
+
+                    if (!telnetSTA) {
+                        DEBUG_MSG_P(PSTR("[TELNET] Rejecting - Only local connections\n"));
+                        _telnetDisconnect(i);
+                        return;
+                    }
+                }
+
+                _telnetNotifyConnected(i);
+
+                break;
+            }
+        }
 
+        //no free/disconnected spot so reject
+        if (i == TELNET_MAX_CLIENTS) {
+            DEBUG_MSG_P(PSTR("[TELNET] Rejecting - Too many connections\n"));
+        }
+    }
+
+    // Read data from clients
+    for (int i = 0; i < TELNET_MAX_CLIENTS; i++) {
+        while (_telnetClients[i] && _telnetClients[i]->available()) {
```

Comment:
One last thing though. What happens when sync client times out? (network connection is cut) Should it check for ->connected() or new client will replace the old one?

Reply:
Hmm I didn't think about that one. Best would probably be to check ``connected()`` and free the pointer at a timeout.

Reply:
Fixed in https://github.com/xoseperez/espurna/pull/1799/commits/da065b2b6e00bf9ecec0105327fe3037db6e490e

Reply:
Think so. But I don't see any ack timeout / keepalive mechanism in ClientContext or WiFiClient, to prevent the thing from #1759. I don't know, does telnet support any protocol-level keepalive? Async variant does ack timeout, it will fail after some writes